src/parallel/HighsTaskExecutor.h

```diff
@@ -85,47 +90,75 @@ class HighsTaskExecutor {
-      HighsSpinMutex::yieldProcessor();
-    }
-    // now acquire a reference count of the global executor
-    threadLocalExecutorHandle() = *executor;
-    HighsSplitDeque* localDeque = workerDeques[workerId].get();
-    threadLocalWorkerDeque() = localDeque;
-    HighsTask* currentTask = workerBunk->waitForNewTask(localDeque);
-    while (currentTask != nullptr) {
-      localDeque->runStolenTask(currentTask);
-
-      currentTask = random_steal_loop(localDeque);
-      if (currentTask != nullptr) continue;
-
-      currentTask = workerBunk->waitForNewTask(localDeque);
+  static void run_worker(int workerId, HighsTaskExecutor* ptr) {
+    auto& executorHandle = threadLocalExecutorHandle();
+    executorHandle.ptr = ptr;
+
+    // check if main thread has shutdown before thread has started
+    // if (ptr->mainWorkerId.load() != std::thread::id()) {
+    if (!(executorHandle.isMain)) {
+      HighsSplitDeque* localDeque = ptr->workerDeques[workerId].get();
+      threadLocalWorkerDeque() = localDeque;
+
+      HighsTask* currentTask = ptr->workerBunk->waitForNewTask(localDeque);
+      while (currentTask != nullptr) {
+        localDeque->runStolenTask(currentTask);
+
+        currentTask = ptr->random_steal_loop(localDeque);
+        if (currentTask != nullptr) continue;
+
+        currentTask = ptr->workerBunk->waitForNewTask(localDeque);
+      }
     }
+
+    threadLocalExecutorHandle().dispose();
   }
 
  public:
   HighsTaskExecutor(int numThreads) {
     assert(numThreads > 0);
-    mainWorkerHandle.store(nullptr, std::memory_order_relaxed);
+
     workerDeques.resize(numThreads);
     workerBunk = cache_aligned::make_shared<HighsSplitDeque::WorkerBunk>();
     for (int i = 0; i < numThreads; ++i)
       workerDeques[i] = cache_aligned::make_unique<HighsSplitDeque>(
           workerBunk, workerDeques.data(), i, numThreads);
 
     threadLocalWorkerDeque() = workerDeques[0].get();
-    for (int i = 1; i < numThreads; ++i)
-      std::thread([&](int id) { run_worker(id); }, i).detach();
+    workerThreads.reserve(numThreads - 1);
+    referenceCount.store(numThreads);
+
+    for (int i = 1, numThreads = workerDeques.size(); i < numThreads; ++i) {
+      workerThreads.emplace_back(
+          std::move(std::thread(&HighsTaskExecutor::run_worker, i, this)));
+    }
+  }
+
+  void stopWorkerThreads(bool blocking = false) {
+    // auto id = mainWorkerId.exchange(std::thread::id());
+    // if (id == std::thread::id()) return;  // already been called
+    auto& executorHandle = threadLocalExecutorHandle();
+    if (executorHandle.isStopped) return;
+
+    // now inject the null task as termination signal to every worker
+    for (auto& workerDeque : workerDeques) {
+      workerDeque->injectTaskAndNotify(nullptr);
+    }
+
+    // only block if called on main thread, otherwise deadlock may occur
+    // if (blocking && std::this_thread::get_id() == id) {
+    // threadLocalExecutorHandle();
+    if (blocking && executorHandle.isStopped) {
```

Comment:
Need to check `isMain` instead of `isStopped`.  If we are within a worker thread the call to `.join()` doesn't make sense, i.e., we are joining on ourselves?

According to docs it would throw `resource_deadlock_would_occur` if `workerThread.get_id() == std::this_thread::get_id()` (deadlock detected).